cloudferrylib/os/actions/transport_compute_resources.py

```diff
@@ -38,3 +40,144 @@ def run(self, info=None, identity_info=None, **kwargs):
+    | deleted     | int(11)      | YES  |     | NULL    |                |
+    +-------------+--------------+------+-----+---------+----------------+
+
+    # Migration process notes
+
+     - It is assumed that user names in keystone are unique
+     - It is assumed that users are migrated to DST
+     - Since keystone backend can be LDAP as well, user names and IDs must be
+       fetched through keystone API, not the low-level SQL
+     - Update DST DB with key pairs from SRC using SQL
+    """
+
+    @staticmethod
+    def _check_users(dst_users, src_users):
+        src_user_names = set([user.name for user in src_users])
+        dst_user_names = set([user.name for user in dst_users])
+        if not src_user_names.issubset(dst_user_names):
+            LOG.warning("Not all SRC users are present on DST! Absent users: "
+                        "%s", src_user_names - dst_user_names)
+
+    def _get_src_key_pairs(self):
+        src_nova = self.src_cloud.resources[utl.COMPUTE_RESOURCE]
+        key_pairs = map(KeyPair.from_tuple, src_nova.mysql_connector.execute(
+            "SELECT * FROM nova.key_pairs;").fetchall())
+        return key_pairs
+
+    def _replace_src_user_ids_with_dst(self, src_users, dst_users, key_pairs):
+        """Since user IDs on SRC and DST may be different, we must update the
+        DST ids for key pairs with IDs from DST (see `nova.key_pairs` table
+        schema)"""
+        def find(predicate, l):
+            for item in l:
+                if predicate(item):
+                    return item
+
+        dst_nova = self.dst_cloud.resources[utl.COMPUTE_RESOURCE]
+        for key_pair in key_pairs:
+            src_user = find(lambda u: u.id == key_pair.user_id, src_users)
+            if src_user is None:
+                # In case user was deleted from keystone, but key pair is
+                # still there - ignore the key
+                continue
+
+            dst_user = find(lambda u: u.name == src_user.name, dst_users)
+
+            key_pair.user_id = dst_user.id
+
+            LOG.debug('Adding key pair %s for user %s on DST', key_pair.name,
+                      src_user.name)
+
+            key_pair_dict = key_pair.to_dict()
+            kp_fields = str(key_pair_dict.keys()).translate(None, "[]'")
+            kp_values = ','.join(['"{}"'.format(kp) for kp in key_pair_dict.values()])
+            sql = ("INSERT INTO nova.key_pairs ({fields}) "
+                   "VALUES ({values}) ON DUPLICATE KEY UPDATE id=id;"
+                   ).format(fields=kp_fields, values=kp_values)
+            dst_nova.mysql_connector.execute(sql)
+
+    def run(self, **kwargs):
+        src_keystone = self.src_cloud.resources[utl.IDENTITY_RESOURCE]
+        dst_keystone = self.dst_cloud.resources[utl.IDENTITY_RESOURCE]
+
+        LOG.info('Fetching users present on SRC and DST, this may take awhile')
+
+        src_users = src_keystone.get_users_list()
+        dst_users = dst_keystone.get_users_list()
```

Comment:
It's definitely a very bad idea to fetch all users, there can be hundreds of thousands of them in corporate LDAP in general case. I will rewrite this to only get user ID which has key pair on SRC